exec.go

```diff
@@ -82,35 +79,37 @@ func (db *DB) Insert(table string, data interface{}) error {
 		return err
 	}
 
-	if !structMode {
-		for i := 0; i < rv.Len(); i++ {
-			row := reflect.Indirect(rv.Index(i))
-			insert_id, structInfo, err := db.insertStruct(table, row.Interface())
-			if err != nil {
-				return err
-			}
-			pk := structInfo.onlyPrimaryKey()
-			if pk != nil && pk.structField.Type.Kind() == reflect.Int64 {
-				setPrimaryKey(row.FieldByName(pk.name), insert_id)
-			}
-		}
-	} else {
-		insert_id, structInfo, err := db.insertStruct(table, rv.Interface())
+	insert := func(val reflect.Value) error {
+		insert_id, structInfo, err := db.insertStruct(table, val.Interface())
 		if err != nil {
 			return err
 		}
 		pk := structInfo.onlyPrimaryKey()
-		// log.Printf("PK: %d", insert_id)
 		if pk != nil && pk.structField.Type.Kind() == reflect.Int64 {
-			setPrimaryKey(rv.FieldByName(pk.name), insert_id)
+			setPrimaryKey(val.FieldByName(pk.name), insert_id)
 		}
+		return nil
+	}
+
+	if structMode {
+		return insert(rv)
 	}
 
+	for i := 0; i < rv.Len(); i++ {
+		row := reflect.Indirect(rv.Index(i))
+		err = insert(row)
+		if err != nil {
+			return err
+		}
+	}
 	// data
 	return nil
 }
 
 func setPrimaryKey(rv reflect.Value, id int64) {
+	if !rv.CanAddr() {
```

Comment:
I would leave this to the caller of the function. Here, its better to panic. We already do additional checks before setPrimaryKey on the caller site.